packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,888 @@
+  /// [RestorationId]s will trigger in debug mode.
+  ///
+  /// No-op if the provided bucket is already a child of this bucket.
+  void adoptChild(RestorationBucket child) {
+    assert(child != null);
+    if (child._parent != this) {
+      child._parent?._removeChildData(child);
+      child._parent = this;
+      _addChildData(child);
+      if (child._manager != _manager) {
+        _recursivelyUpdateManager(child);
+      }
+    }
+    assert(child._parent == this);
+    assert(child._manager == _manager);
+  }
+
+  void _dropChild(RestorationBucket child) {
+    assert(child != null);
+    assert(child._parent == this);
+    _removeChildData(child);
+    child._parent = null;
+    if (child._manager != null) {
+      child._manager = null;
+      _recursivelyUpdateManager(child);
+    }
+  }
+
+  void _recursivelyUpdateManager(RestorationBucket bucket) {
+    bucket._manager = _manager;
+    bucket._visitChildren(_recursivelyUpdateManager);
+  }
+
+  void _finalize() {
+    _childrenToAdd.forEach((RestorationId id, Set<RestorationBucket> buckets) {
+      assert(() {
+        final int claimCount = buckets.length + (_claimedChildren.containsKey(id) ? 1 : 0);
+        if (claimCount > 1) {
+          throw FlutterError.fromParts(<DiagnosticsNode>[
+            ErrorSummary('Multiple owners claimed child RestorationBuckets with the same ID.'),
+            ErrorDescription('The following owners claimed child RestorationBuckets with id "$id" from the parent $this:'),
+            ...buckets.map((RestorationBucket bucket) => ErrorDescription(
+              ' * ${bucket.debugOwner}',
+            )),
+            if (_claimedChildren.containsKey(id))
+              ErrorDescription(' * ${_claimedChildren[id].debugOwner} (current owner)'),
+          ]);
+        }
+        return true;
+      }());
+      _finalizeAddChildData(buckets.first);
+    });
+    _childrenToAdd.clear();
+  }
+
+  void _removeChildData(RestorationBucket child) {
+    assert(child != null);
+    assert(child._parent == this);
+    if (_claimedChildren.remove(child.id) == child) {
+      _rawChildren.remove(child.id.value);
+      if (_rawChildren.isEmpty) {
+        _rawData.remove(_childrenMapKey);
+      }
+      _manager?.scheduleSerialization();
+      return;
+    }
+    _childrenToAdd[child.id]?.remove(child);
+    if (_childrenToAdd[child.id]?.isEmpty == true) {
+      _childrenToAdd.remove(child.id);
+    }
+  }
+
+  void _addChildData(RestorationBucket child) {
+    assert(child != null);
+    assert(child._parent == this);
+    if (_claimedChildren.containsKey(child.id)) {
+      // Delay addition until the end of the frame in the hopes that the current
+      // owner of the child with the same id will have given up that child by
+      // then.
+      _childrenToAdd.putIfAbsent(child.id, () => <RestorationBucket>{}).add(child);
+      _manager.scheduleSerialization(finalizer: _finalize);
```

Comment:
Elsewhere you're allowing manager to be null, but not here. Maybe add an assert that it's not, or use the `?` again.

Similar question to the public methods though - what happens if it is null here? Seems like bad news.